### DISCUSSION TITLE
Recycling smarter

### DISCUSSION THREAD

**Discussion by DGolubets** [bikeshedder/deadpool#231] [Ideas]
At the moment Deadpool recycles a connection every time it gets an existing one from the pool.
This introduces additional latency.
I wonder if recycling can be smarter and only check connections which were idle for a some time?

**Comment by bikeshedder** [comment]
You can add a `pre_recycle` hook and discard objects there. The hooks API isn't very pretty at the moment due to #200 but it's quite functional as shown in the tests:

https://github.com/bikeshedder/deadpool/blob/f35ac2a93657fc5a17aa53389b7baf3239123d3c/tests/managed_hooks.rs#L131-L141

That `metrics` parameters gives you access to the object [Metrics](https://docs.rs/deadpool/latest/deadpool/managed/struct.Metrics.html)

e.g. if you wanted to discard objects which weren't recycled within the last minute or so you can do it like that:
```rust
     .pre_recycle(Hook::sync_fn(|_, metrics| { 
         if metrics.last_used() > Duration::from_secs(60) {
             Err(HookError::Continue(None)) 
         } else { 
             Ok(()) 
         } 
     })) 
```

**Comment by bikeshedder** [comment]
Sorry. I got that wrong. You want it the other way around. Rather than fast discarding possibly broken objects you don't want to run the recycle method at all for objects which have just been returned to the pool.

**Comment by DGolubets** [comment]
Maybe `pre_recycle` hook can return a `bool` flag telling whether to recycle or not?

**Reply by bikeshedder** [reply]
(deleted some misleading comments... once again I got that wrong...)

**Reply by bikeshedder** [reply]
As a first step we could pass the `Metrics` object into the `recycle` method of the pool. That way it would be possible to write a wrapper Manager that does exactly that.

I just opened an issue about that:

- #236

**Comment by bikeshedder** [comment]
This is a requirement for #167 and will be tracked as part of this issue:

- #167
